### PR TITLE
WIP: pixel classifier

### PR DESCRIPTION
**Pull request by LeeKamentsky**
The idea here is to use CellProfiler to train a classifier by feeding it objects representing ground truth and gray-scale images representing the channels. Underlying, the data structure is 5D, but only channel, I and J are used.

A sampling window is taken around each pixel to create a feature vector. Dimensionality reduction is done via PCA to get a set of eigentextures. The output of these eigentexture filters creates a reduced feature vector for a classifier, e.g. random forests. Currently, things are trained using randomly selected pixels and then two rounds of training are done with random pixels augmented with misclassified pixels.

Classification applies the kernel to the image (in 128x128 chunks), runs the eigentexture filters to reduce the dimensionality and then classifies using the trained random forest.

### CODE REVIEW

**Comment by LeeKamentsky** [comment]
As of 2/8, I think it's operating correctly, but slowly, especially the training phase. I am creating the per-pixel feature vectors and writing them to disk and it seems that disk transfer is the source of almost all of the time used to train. For any reasonable training set, the number of samples drawn is far less (10% or fewer) than the number of examples in the ground truth, so it may be cheaper to store the images and create the feature vectors for the training set on the fly (compute time << disk access time).

I am in the middle of a run using 6 images from the Moy C. elegans screen and in the initial pass, 97% of the background pixels were classified correctly. There are two further rounds, each of which augment the training set from the misclassified pixels, so the final will be much better.

**Comment by LeeKamentsky** [comment]
My first classified image. No illumination correction, training done on another plate.

first_light

**Comment by 0x00B1** [comment]
@holgerhennig

**Comment by 0x00B1** [comment]
@LeeKamentsky can I help with this? Let me know! :)

**Comment by LeeKamentsky** [comment]
You can now use label .tif files exported by Ilastik in addition to objects. I took the liberty of not doing backwards compatibility because this isn't on the master branch yet, so old pipelines using the classify module will not load correctly.

**Comment by 0x00B1** [comment]
> I took the liberty of not doing backwards compatibility because this isn't on the master branch yet, so old pipelines using the classify module will not load correctly.

Could you use an alternative name to mitigate this issue?

**Comment by LeeKamentsky** [comment]
It does use an alternative name - the old one that uses Ilastik is called ClassifyPixels and this one is called Classify. But if David checked out the branch this morning and made a pipeline using the Classify module, it would break. Almost certainly, David is the only person who will suffer from backwards compatibility - just wanted to warn him.

**Comment by 0x00B1** [comment]
Ah, thanks for the clarification!

**Comment by holgerhennig** [comment]
Great, I'll test the pixelbased classifier on a different project, will let you know how it goes

**Comment by 0x00B1** [comment]
It’d be nice if we could depreciate ClassifyPixels entirely. It’d help cleanup `cellprofiler.__main__.py`.

**Comment by holgerhennig** [comment]
Good idea to depreciate ClassifyPixels from a users perspective, it's confusing when there's an old classifypixels and the new classify

**Comment by 0x00B1** [comment]
@LeeKamentsky Will you upload your example pipeline (i.e. `c:/temp/output/classify/classify.cpproj`)?

File: cellprofiler/modules/classify.py [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+
+        :param filter_bank_name:
+        :param sample:
+
+        :return:
+        """
+        dataset = self.g_filters[filter_bank_name]
+
+        if dataset.attrs[A_CLASS] == CLASS_FILTER:
+            if USE_DOT:
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: `USE_DOT` is always `True`

File: cellprofiler/modules/classify.py [hunk 2]
```diff
@@ -0,0 +1,952 @@
...
+                result = numpy.dot(sample, dataset[:].T)
+            else:
+                #
+                # A dot product... but cluster's np.dot is so XXXXed
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: LOL, is this true?

File: cellprofiler/modules/classify.py [hunk 3]
```diff
@@ -0,0 +1,952 @@
...
+
+        self.path = cellprofiler.settings.DirectoryPath("Classifier folder")
+
+        self.filename = cellprofiler.settings.FilenameText("Classifier file", "Classifier.cpclassifier", get_directory_fn=get_directory_fn,
+                                                           set_directory_fn=set_directory_fn,
+                                                           exts=[("Pixel classifier (*.cpclassifier)", "*.cpclassifier"), ("All files (*.*)", "*.*")])
+
+        self.gt_source = cellprofiler.settings.Choice("Ground truth source", [SRC_OBJECTS, SRC_ILASTIK],
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: `SRC_ILASTIK` is always `"Ilastik"`

File: cellprofiler/modules/classify.py [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+        Set the kernel used to sample pixels from a neighborhood
+
+        :param kernel: an N x 5 matrix of N offsets by 5 dimensions (C, T, Z, Y, X)
+
+        :return:
+        """
+        if DATASET_KERNEL in self.root.keys():
+            del self.root[DATASET_KERNEL]
+
+        self.root.create_dataset(DATASET_KERNEL, data=kernel).attrs[A_CLASS] = CLASS_KERNEL
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: `CLASS_KERNEL` is always `"Kernel"`

File: cellprofiler/modules/classify.py [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+
+        :param classifier_name: save using this name
+        :param sample: S samples x N features
+        :param classes: S class labels indexing into the class names
+        :param algorithm: algorithm to use to train
+
+        :return:
+        """
+        if algorithm is None:
+            algorithm = sklearn.ensemble.ExtraTreesClassifier(n_estimators=N_ESTIMATORS, min_samples_leaf=MIN_SAMPLES_PER_LEAF)
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: `N_ESTIMATORS` is undefined

Comment (0x00B1) [inline cellprofiler/modules/classify.py]: `MIN_SAMPLES_PER_LEAF` is undefined

File: cellprofiler/modules/classify.py [hunk 6]
```diff
@@ -0,0 +1,952 @@
...
+
+        self.min_samples_per_leaf = cellprofiler.settings.Integer("Minimum samples per leaf", DEFAULT_MIN_SAMPLES_PER_LEAF, 1,
+                                                                  doc="This setting determines the minimum number of ground truth pixels that the classifier will use to split a decision tree. There must be at least this number of example pixels in each branch for the classifier to have confidence that the split is real and not just an artifact of an irrelevant measurement.\n Lower this setting if the classifier does a good job on most of the pixels but does not draw sharp distinctions between one class and another at the border between the classes (e.g. at the edges of cells). Raise this setting if the classifier misclassifies pixels that are clearly not the right class - this is overtraining.")
+
+        self.path = cellprofiler.settings.DirectoryPath("Classifier folder")
+
+        self.filename = cellprofiler.settings.FilenameText("Classifier file", "Classifier.cpclassifier", get_directory_fn=get_directory_fn,
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: Can you use lambda rather `get_directory_fn`? What about `set_directory_fn`?

Comment (0x00B1) [inline cellprofiler/modules/classify.py]: Can you use lambda rather `get_directory_fn`? What about `set_directory_fn`?

File: cellprofiler/modules/classify.py (hunk 5 above)
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: Shouldn’t this be the default argument for `algorithm` (rather than “None”)?

File: cellprofiler/modules/classify.py [hunk 7]
```diff
@@ -0,0 +1,949 @@
...
+                result += group.visible_settings()
+
+            result.append(self.add_output_button)
+
+        return result
+
+    def prepare_settings(self, settings):
+        for count, sequence, add_fn in zip([int(_) for _ in settings[:4]], (self.object_classes, self.label_classes, self.images, self.outputs),
+                                           (self.add_objects, self.add_labels, self.add_image, self.add_output)):
+            del sequence[:]
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: Why the explicit memory de-allocation?

File: cellprofiler/modules/classify.py [hunk 8]
```diff
@@ -0,0 +1,949 @@
...
+        return self.g_training_set[class_name]
+
+    @property
+    def ground_truth_chunk_size(self):
+        """The size of a chunk of ground truth that fits in memory"""
+        return int(50 * 1000 * 1000 / self.get_kernel().shape[0])
+
+    @property
+    def pix_chunk_size(self):
+        return 1000 * 1000
```
Comment (0x00B1) [inline cellprofiler/modules/classify.py]: ???

File: cellprofiler/modules/classify.py (hunk 2 above)
Comment (LeeKamentsky) [inline cellprofiler/modules/classify.py]: feel free to change USE_DOT and find out

File: cellprofiler/modules/classify.py (hunk 7 above)
Comment (LeeKamentsky) [inline cellprofiler/modules/classify.py]: It's not a deallocation, it's deleting the elements of the list.

File: cellprofiler/modules/classify.py (hunk 6 above)
Comment (LeeKamentsky) [inline cellprofiler/modules/classify.py]: For get_directory_fn, a lambda function is possible, for set_directory_fn, it's a stretch and only possible because of the order of outputs from self.path.get_parts_from_path.

    set_directory_fn=lambda path: self.path.join_parts(*self.path.get_parts_from_path(path))

File: cellprofiler/modules/classify.py (hunk 5 above)
Comment (LeeKamentsky) [inline cellprofiler/modules/classify.py]: Good catch on N_ESTIMATORS and MIN_SAMPLES_PER_LEAF. That's a legacy of me fooling around with the classifier when those were static variables.

**Comment by LeeKamentsky** [comment]
Projects that train and classify (with not so good ground truth) are at /imaging/cold/People/Lee/wormclassifier

**Comment by holgerhennig** [comment]
Testing the module Classify (pixel-based classifier) with Myelin sheaths images. This is more of a tough example to test with (not necessarily expecting great results w/ a pixel-based classifier). Biological aim is to differentiate between complete (marked green) and incomplete tubes (marked red/purple) in image

General comments/recommendations
- module classify works great! 2 basic steps: train, classify
- improve performance: currently several minutes (each for training and classifying)
- training mode: Classifier.cpclassifier file is huge (~1G) compared to input image file (7KB)
- so far: PCA features are calculated by the classify module. Add other features (texture etc)? 
- As a guide for users to know when the cpclassifier file is either being created (train mode) or read (classify mode), I'd recommend: In train mode, replace the text "Classifier file" by "Output classifier file". In classify mode, replace the text "Classifier file" by "Input classifier file"
- class label names need to be inserted manually (not read from file), but is no big deal when having only few classes

General workflow involving labeling in Ilastik
1) label the classes (here we have 4 classes: complete, incomplete, background, junk) in Ilastik and export labels as tiff, e.g. labels.tif
2) in CP's module classify in train mode, choose Ilastik as ground truth input and point to labels.tif. Train the classifier
Note: A separate CP pipeline for training and classifying needs to be built
3) Build pipeline using classify module in classify mode, this generates the probability maps. Then identify primary objects (complete, incomplete myelin sheaths) based on the probability maps

Results of CP pixel-based classifier:
overlay008

For comparison: labeling of a biologist (not used for training the classifier)
max_021616expt series008 -mbp example annotated

**Comment by LeeKamentsky** [comment]
@holgerhennig 
Performance: there are two factors that determine performance during training. The first is the radius and a 2x increase in radius increases the processing time and memory by 4x. As you get further away from the pixel of interest, there is less correlation between a pixel and the  one of interest, so you could use a strategy that produces fewer pixel features at larger radii for input to PCA. For instance, you could use bicubic spline interpolation to sample the annulus from radius / 2 to radius at 1/2 the resolution.
The second factor that influences performance during training (but not classification) is the size and number of rounds of training. I subsample from the ground-truth data currently and run three rounds to try to zero in on the pixels that are at the error margins of the classifier. So you could reduce the training time linearly with the # of rounds or the # of pixels per round.

Classifier file size: I use the classifier file as temp storage during training. There's no way to shrink the size of an HDF file, so all that temp storage remains after the file has been used. It would be pretty straightforward to copy only the datasets that are needed for classification - these are pretty small.

Adding features: I wish I had a tool that would let you visualize the eigentextures - when you have ~100 of them, they are very comprehensive. Perhaps something like eigenvalues of the hessian might be useful because it's not a convolution. Perhaps just using SIFT or SURF instead of PCA would work too. But my gut is that adding features will not improve performance noticeably.

Class labels: these are not available from the .tif files output from Ilastik (yes, I looked through the TIF metadata). Possibly they are from the .HDF file and they definitely are available if you use CellProfiler .tif object files.

**Comment by bethac07** [comment]
Has anyone successfully gotten this working using the "Object" based classification?  I've been trying to use it and after training it in four slightly different ways (using a full image set, using a subset of the images, using all of the channels, using only one of the channels), each one produced a probability map that was just pure white.  :-/

**Comment by dlogan** [comment]
Not me.  @holgerhennig did though.  Have you tried looking at the raw data, e.g. in Matlab?  Maybe the values (erroneously) range from [1,N] and thus look white?

**Comment by bethac07** [comment]
Thanks @dlogan , I saw @holgerhennig had done it using ilastik as an input but didn't know he'd gotten it working with objects as well.  

I haven't tried that but I'll try it next; I did mouse over it and it was a sea of 1.00000 but I'll export it and take a look.

**Comment by mcquin** [comment]
Is anyone committed to finishing this work? @LeeKamentsky @0x00b1

**Comment by AnneCarpenter** [comment]
There IS a plan to implement pixel classification via a better connection to ilastik #2393 - that work might make this less necessary.
